signer/src/transaction_signer.rs

```diff
@@ -256,6 +285,87 @@ where
         Ok(true)
     }
 
+    #[tracing::instrument(skip(self))]
+    async fn handle_wsts_message(
+        &mut self,
+        msg: &message::WstsMessage,
+        bitcoin_chain_tip: &model::BitcoinBlockHash,
+    ) -> Result<(), error::Error> {
+        tracing::info!("handling message");
+        match &msg.inner {
+            wsts::net::Message::DkgBegin(_) => {
+                let state_machine = self.create_new_state_machine()?;
+                self.wsts_state_machines.insert(msg.txid, state_machine);
+                self.relay_message(msg.txid, &msg.inner, bitcoin_chain_tip)
+                    .await?;
+            }
+            wsts::net::Message::DkgPublicShares(_) => {
+                self.relay_message(msg.txid, &msg.inner, bitcoin_chain_tip)
+                    .await?;
+            }
+            wsts::net::Message::DkgPrivateBegin(_) => {
+                self.relay_message(msg.txid, &msg.inner, bitcoin_chain_tip)
+                    .await?;
+            }
+            wsts::net::Message::DkgPrivateShares(_) => {
+                self.relay_message(msg.txid, &msg.inner, bitcoin_chain_tip)
+                    .await?;
+            }
+            wsts::net::Message::DkgEndBegin(_) => {
+                self.relay_message(msg.txid, &msg.inner, bitcoin_chain_tip)
+                    .await?;
+                self.store_dkg_shares(&msg.txid).await?;
+            }
+            wsts::net::Message::NonceRequest(_) => {
+                // TODO(296): Validate that message is the appropriate sighash
+                self.relay_message(msg.txid, &msg.inner, bitcoin_chain_tip)
+                    .await?;
+            }
+            wsts::net::Message::SignatureShareRequest(_) => {
+                // TODO(296): Validate that message is the appropriate sighash
+                self.relay_message(msg.txid, &msg.inner, bitcoin_chain_tip)
+                    .await?;
+            }
+            _ => {
+                tracing::debug!("ignoring message");
+            }
+        }
+
+        Ok(())
+    }
+
+    #[tracing::instrument(skip(self))]
+    async fn relay_message(
+        &mut self,
+        txid: bitcoin::Txid,
+        msg: &wsts::net::Message,
+        bitcoin_chain_tip: &model::BitcoinBlockHash,
+    ) -> Result<(), error::Error> {
+        let Some(state_machine) = self.wsts_state_machines.get_mut(&txid) else {
+            tracing::warn!("missing signing round");
+            return Ok(());
+        };
+
+        let outbound_messages = state_machine.process(msg).map_err(error::Error::WSTS)?;
+
+        for outbound_message in outbound_messages.iter() {
+            // The WSTS state machine assume we read our own messages
+            state_machine
+                .process(outbound_message)
+                .map_err(error::Error::WSTS)?;
```

Comment:
Oh, this is very important. Good to know.

Reply:
Yeah, I wouldn't have discovered it if it weren't for the unit tests. I'm so happy we have the ability to run this component in isolation and discover these quirks early :)